Additional support for factur-x (PDF-A3)

DkDavid opened the pull request:
The factur-x (ZUGFeRD) standard needs the attached files listed in the AF-tag.
Furthermore, additional metadata is necessary for the embedded files.

Easy solution: Allow users to supply additionalParams.

To access the attached files of factur-x invoices, access to the PDFRawStream is needed.

DkDavid committed:
Added support for factur-x, additional metadata for embedded files, add embedded files to AF tag, make PDFRawStram available.

DkDavid committed:
Add PDFRawStram to lookup function

Hopding commented:
Hello @DkDavid!

I'm not familiar with the details of the ZUGFeRD standard. I don't know anything about the `AF` entry and the PDF spec doesn't seem to list it in the table of catalog entries. Can you please point me to some documentation about this? I'd like to understand it better before merging this.

DkDavid commented:
Hi Hopding,

sorry for my late reply.
The standard is used to attach invoice meta-data to PDF/A3 documents (Germany: ZUGFeRD, France: Factur-X, since version 2.1 the name is Factur-X and a European standard exists). Most additions can easily be made using the existing API. I can add a minimal example to get the conformance with the standard, if desired. However, additional params for the embedded file as well as the AF entry are necessary. 

The specification can be downloaded here: https://www.ferd-net.de/standards/zugferd-versionsarchiv/zugferd-2.1.html
The additional AF entry is mentioned in "ZUGFeRD-2.1.0 - Spezifikation_TA_Anlage-A" - page 8 (diagram) for example.

DkDavid committed:
Added CIDToGIDMap to achieve PDF-A3 compliance

Hopding commented:
@DkDavid Would it be possible to add specific options to the embedded file options? I would prefer that approach as it is easier to document and use than allowing users to pass arbitrary values. Note that I haven't had a chance to familiarize myself with the ZUGFeRD specification, so just let me know if there's a specific reason this wouldn't work.

DkDavid commented:
@Hopding Yes, that would be possible. The "additional param" needed for factur-x is: additionalParams: {'AFRelationship': 'Alternative'} for the xml attachment. 
I just thought it would be more flexible for future standards. However, the optional param "AFRelationship" would be sufficient for factur-x.  

This as well as the proposed change in PDFEmbeddedFile.ts, and the addition of the CIDToGIDMap would be necessary to fulfill compliance. Everything else is easily possible with the current api.  
For CIDToGIDMap see https://github.com/Hopding/pdf-lib/issues/230#issuecomment-705460408

Hopding commented on src/core/embedders/FileEmbedder.ts:
```diff
@@ -2,12 +2,14 @@ import PDFString from 'src/core/objects/PDFString';
   description?: string;
   creationDate?: Date;
   modificationDate?: Date;
+  additionalParams?: LiteralObject;
```
Per https://github.com/Hopding/pdf-lib/pull/508#issuecomment-719968177 can we change this to:
```ts
/** 
 * From the ZUGFeRD specification, Supplement-A, section **2.2.2. Data Relationship**.
 * See:
 * * https://www.ferd-net.de/standards/zugferd-versionsarchiv/zugferd-2.1.html
 * * http://www.awv-net.de/updates/zugferd21/zugferd21en.zip
 */
export enum AFRelationship {
  Data = 'Data',
  Source = 'Source',
  Alternative = 'Alternative',
  Supplement = 'Supplement',
  Unspecified = 'Unspecified'
}

export interface EmbeddedFileOptions { 
  ...,
  afRelationship: AFRelationship;
}
```

Hopding commented on src/core/embedders/FileEmbedder.ts:
```diff
@@ -62,6 +66,7 @@ class FileEmbedder {
       UF: PDFHexString.fromText(this.fileName),
       EF: { F: embeddedFileStreamRef },
       Desc: description ? PDFHexString.fromText(description) : undefined,
+      ...additionalParams
```
See previous comment. Can we change this to:
```ts
AFRelationship: afRelationship ?? undefined,
```

Hopding commented on src/core/objects/PDFArray.ts:
```diff
@@ -11,6 +11,7 @@ import PDFString from 'src/core/objects/PDFString';
 import PDFContext from 'src/core/PDFContext';
 import CharCodes from 'src/core/syntax/CharCodes';
 import { PDFArrayIsNotRectangleError } from '../errors';
+import PDFRawStream from './PDFRawStream';
```
```suggestion
import PDFRawStream from 'src/core/objects/PDFRawStream';
```

Hopding commented on src/core/PDFContext.ts:
```diff
@@ -21,15 +21,15 @@ import { typedArrayFor } from 'src/utils';
 type LookupKey = PDFRef | PDFObject | undefined;
 
-interface LiteralObject {
+export interface LiteralObject {
```
Per previous comments we should no longer need to export these.

Hopding reviewed (changes requested):
@DkDavid Thanks for working on this! I certainly appreciate your desire to make the API extensible. It might alleviate the need to add specific new options in the future. However, I'm concerned that it could also result in a more obscure and less well-documented API that would be more difficult to understand, use, and maintain. So I've requested some changes to implement this feature more in-line with existing API options.

In addition to my other comments, we'll need to add an assertion to the following method for the new option it accepts:
https://github.com/Hopding/pdf-lib/blob/30d2aa22c0c0d694189ae3202562d4c0565cce42/src/api/PDFDocument.ts#L810-L822

The assertion should be:
```
assertIsOneOfOrUndefined(options.afRelationship, 'options.afRelationship', AFRelationship);
```
(see [this assertion](https://github.com/Hopding/pdf-lib/blob/30d2aa22c0c0d694189ae3202562d4c0565cce42/src/api/PDFPage.ts#L894) for reference).

Commit:
implement requested changes, update reference to PDF-A3

Commit:
cleanup from previous implementation

DkDavid commented:
@Hopding I digged into the Factur-X (european harmonized name of ZugFERD) and the PDF-A3 specification. Nearly all of the changes can be led back to the PDF-A3 specification. Therefore I changed the references accordingly and implemented your comments. 

To reach PDF-A3 compliance three additional things have to be done by the user. All of them are reachable by the current API.

`const hash =   crypto.createHash('sha256').update(DOCUMENTID + new Date().toISOString ).digest(); //crypto.subtle.digest('SHA-512', data);
        const hashArray = Array.from(new Uint8Array(hash));
        const hashHex = hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
        const permanentDocumentId = PDFHexString.of(hashHex);
        const changingDocumentID = permanentDocumentId; // updated when document is changed 
        doc.context.trailerInfo.ID = doc.context.obj([permanentDocumentId, changing]);

       // MarkInfo specifies if the document is a tagged pdf file. => 9.7 https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/pdf_reference_archives/PDFReference.pdf
        const rootref = doc.context.obj({Marked: true})
        doc.catalog.set(PDFName.of('MarkInfo'), rootref);
      
       // => 14.7.2 https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf
        const structTreedata = doc.context.obj({Type: PDFName.of('StructTreeRoot')})
        const structTreeref = doc.context.register(structTreedata);
        doc.catalog.set(PDFName.of('StructTreeRoot'), structTreeref);`

Hopding reviewed (approved):
@DkDavid This is perfect! Thanks for working on this and figuring out what is necessary to achieve PDF-A3 compliance. This feature will go out in the next release of pdf-lib 🎉

Hopding commented:
The changes in this PR were just released in `v1.13.0`.

ram-you commented:
Hello everyone.
First of all, I apologize for commenting on a closed issue.
I wanted to ask if it is possible to provide us with a working example to generate a PDF with the Factur-X (european harmonized name of ZugFERD) and the PDF-A3 specification.
Thanks in advance and sorry for the inconvenience.

kemotx90 commented:
> Hello everyone. First of all, I apologize for commenting on a closed issue. I wanted to ask if it is possible to provide us with a working example to generate a PDF with the Factur-X (european harmonized name of ZugFERD) and the PDF-A3 specification. Thanks in advance and sorry for the inconvenience.

for PDF-A3 follow this #230
